pep795.md

```diff
@@ -0,0 +1,1938 @@
+  to declarations.
+- Freezing a module object propagates to any object
+  reachable from the module object, directly or indirectly,
+  but not declarations.
+
+**Freezing rules:**
+
+- Objects support freezing by default, but note that only
+  objects whose types are defined to be immutable by this
+  PEP can be frozen.
+- An object can opt out of support for freezing by setting
+  `__freezable__ = 0`.
+
+**Immutability invariant:**
+
+- An immutable object only references other immutable
+  objects.
+
+
+## What we propose (Alternative 2)
+
+The second alternative takes the most permissive model
+as its starting point, but introduces limitations on the
+propagation of freezing. This means that freezing objects
+becomes a little more involved since freezing may involve
+multiple steps, but also limits the freeze propagation.
+In summary, we distinguish between freezing "objects"
+and "declarations" (e.g. types, functions, methods, and
+modules), and try to limit propagation based on these
+distinctions:
+
+**Propagation rules:**
+
+- Freezing an object propagates to other objects,
+  but not to declarations.
+- Freezing a function or method propagates to other
+  functions or methods.
+- Freezing a type propagates to its super types and
+  functions in the type and supertypes.
+- Freezing a module object propagates to any object
+  or declaration reachable from the module object,
+  directly or indirectly. (Notably, this may include
+  submodules.)
+- Freezing a declaration decorated with @freezable
+  may propagate to objects (causing them to be frozen).
+
+**Freezing rules:**
+
+- Module objects must opt-in to support freezing.
+- Freezing a declaration inside a module that does not
+  support freezing will fail unless the declaration is
+  decorated with @freezable.
+- Freezing a declaration that captures mutable state will fail
+  unless the declaration is decorated with @freezable.
+- Objects "inherit" their support for freezing from its
+  class. This can be overridden.
+
+**Immutability invariant:**
+
+- An immutable object only references other immutable
+  objects, with the exception of mutable objects that
+  preserve subinterpreter isolation but are nevertheless
+  marked as immutable in the system. (For example our escape
+  hatches.)
+
+Defining a type or function that is frozen right after
+creation is possible using the `@frozen` decorator.
+
+See below for additional details and examples.
+
+
+### Freezing objects
+
+An object can be frozen if it satisfies *all* of the
+constraints below:
+
+- It does not explicitly opt-out of support for freezing
+- Its class is immutable
+- All the objects reachable from the object can be frozen
+
+Failing to satisfy any of the above constraints will
+cause freezing to fail, leaving all objects in their
+original state.
+
+Opting out of freezable can be done through a field
+`__freezable__` that holds any other value than `1`,
+or by using an `@freezable` decorator on a declaration.
```

Comment:
Is this really opting out? I thought the declaration opts in?

Do you mean `unfreezable`?